internal/service/dynamodb/global_secondary_index.go

```diff
@@ -0,0 +1,934 @@
+				)
+
+				return
+			}
+		} else {
+			ut.AttributeDefinitions = append(ut.AttributeDefinitions, ddbtypes.AttributeDefinition{
+				AttributeName: aws.String(rangeKey),
+				AttributeType: data.RangeKeyType.ValueEnum(),
+			})
+		}
+
+		keySchema = append(keySchema, ddbtypes.KeySchemaElement{
+			AttributeName: aws.String(rangeKey),
+			KeyType:       ddbtypes.KeyTypeRange,
+		})
+	}
+
+	projection := &ddbtypes.Projection{
+		ProjectionType: data.ProjectionType.ValueEnum(),
+	}
+
+	if !data.NonKeyAttributes.IsNull() && !data.NonKeyAttributes.IsUnknown() {
+		response.Diagnostics.Append(
+			data.NonKeyAttributes.ElementsAs(ctx, &projection.NonKeyAttributes, false)...,
+		)
+	}
+
+	action := &ddbtypes.CreateGlobalSecondaryIndexAction{
+		IndexName:  data.Name.ValueStringPointer(),
+		KeySchema:  keySchema,
+		Projection: projection,
+	}
+
+	if data.ReadCapacity.IsNull() || data.ReadCapacity.IsUnknown() {
+		data.ReadCapacity = types.Int64Value(0)
+	}
+
+	if data.WriteCapacity.IsNull() || data.WriteCapacity.IsUnknown() {
+		data.WriteCapacity = types.Int64Value(0)
+	}
+
+	billingMode := ddbtypes.BillingModeProvisioned
+	if table.BillingModeSummary != nil {
+		billingMode = table.BillingModeSummary.BillingMode
+	}
+
+	if billingMode == ddbtypes.BillingModeProvisioned {
+		rc := int(data.ReadCapacity.ValueInt64())
+		wc := int(data.WriteCapacity.ValueInt64())
+
+		if rc == 0 && table.ProvisionedThroughput != nil {
+			rc = int(aws.ToInt64(table.ProvisionedThroughput.ReadCapacityUnits))
+		}
+		if wc == 0 && table.ProvisionedThroughput != nil {
+			wc = int(aws.ToInt64(table.ProvisionedThroughput.WriteCapacityUnits))
+		}
+
+		provisionedThroughputData := map[string]any{
+			"read_capacity":  rc,
+			"write_capacity": wc,
+		}
+		action.ProvisionedThroughput = expandProvisionedThroughput(provisionedThroughputData, billingMode)
+	} else if !data.OnDemandThroughputs.IsNull() && !data.OnDemandThroughputs.IsUnknown() {
+		var odts []onDemandThroughputModel
+		response.Diagnostics.Append(data.OnDemandThroughputs.ElementsAs(ctx, &odts, false)...)
+		if len(odts) > 0 {
+			v := map[string]any{
+				"max_read_request_units":  int(odts[0].MaxReadRequestUnits.ValueInt64()),
+				"max_write_request_units": int(odts[0].MaxWriteRequestUnits.ValueInt64()),
+			}
+			action.OnDemandThroughput = expandOnDemandThroughput(v)
+		}
+	}
+
+	var wts []warmThroughputModel
+	response.Diagnostics.Append(data.WarmThroughputs.ElementsAs(ctx, &wts, false)...)
+	if len(wts) > 0 {
+		action.WarmThroughput = expandWarmThroughput(map[string]any{
+			"read_units_per_second":  int(wts[0].ReadUnitsPerSecond.ValueInt64()),
+			"write_units_per_second": int(wts[0].WriteUnitsPerSecond.ValueInt64()),
+		})
+	}
+
+	ut.GlobalSecondaryIndexUpdates = []ddbtypes.GlobalSecondaryIndexUpdate{
+		{
+			Create: action,
+		},
+	}
+
+	response.Diagnostics.Append(validateNewGSIAttributes(data, *table)...)
```

Comment:
Let's move validation to after we've retrieved the table but before we start building the input